internal/hcp/internal/controllers/link/controller.go

```diff
@@ -15,17 +18,52 @@ import (
 	pbhcp "github.com/hashicorp/consul/proto-public/pbhcp/v2"
 )
 
-func LinkController(resourceApisEnabled bool, hcpAllowV2ResourceApis bool) *controller.Controller {
+type HCPClientFn func(link *pbhcp.Link) (hcpclient.Client, error)
+
+func NewHCPClient(link *pbhcp.Link) (hcpclient.Client, error) {
+	hcpClient, err := hcpclient.NewClient(config.CloudConfig{
+		ResourceID:   link.ResourceId,
+		ClientID:     link.ClientId,
+		ClientSecret: link.ClientSecret,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return hcpClient, nil
+}
+
+func LinkController(resourceApisEnabled bool, hcpAllowV2ResourceApis bool, hcpClientFn HCPClientFn) *controller.Controller {
 	return controller.NewController("link", pbhcp.LinkType).
 		WithReconciler(&linkReconciler{
 			resourceApisEnabled:    resourceApisEnabled,
 			hcpAllowV2ResourceApis: hcpAllowV2ResourceApis,
+			hcpClientFn:            hcpClientFn,
 		})
 }
 
 type linkReconciler struct {
 	resourceApisEnabled    bool
 	hcpAllowV2ResourceApis bool
+	hcpClientFn            HCPClientFn
+}
+
+func (r *linkReconciler) linkingFailed(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) error {
+	newStatus := &pbresource.Status{
+		ObservedGeneration: res.Generation,
+		Conditions:         []*pbresource.Condition{ConditionFailed},
+	}
+	if resource.EqualStatus(res.Status[StatusKey], newStatus, false) {
```

Comment:
We're doing this status check a lot, so might be useful to have a separate helper method that checks status and writes if changed.

Reply:
Good idea, I added that 👍